README.md
# Eindwerk_Development
Dit is een open source API die ontwikkeld is voor mijn vak development en mijn eindwerk. Op deze website kunnen kinderen via een interactief voorlees verhaal leren over de gevolgen van gezond/ ongezond eten. De open source api is dat je vragen, dankwoorden en technische problemen kan gebruiken in je website. 
Hierdoor hoef je geen lorem ipsum te gebruiken of hoef je geen tijd te steken in mogelijke vragen te bedenken.

# Hoe begin ik er nu aan?
## Lokaal laten draaien
- Je kan de api lokaal laten draaien via Docker. Als je dit niet hebt kan je deze makkelijk downloaden en een account op aanmaken.
- Als dit in orde is hoef je in de terminal enkel docker-compose up in te geven.
- Via een programma zoals Tableplus kan je de database live bekijken.
- Als je index.html opent in jouw browser krijg je de website te zien.

## Endpoints
 - Create: /post/formulier (Via het forumulier kan je vragen, problemen en dankwoorden invoegen)
 - Read: /get/:alleVragen (Je krijgt een overzicht van alle ingezonden forumulieren)
 - Update: /update/:id (Als er een fout is of de gegevens van de persoon moeten aangepast worden kan dit via deze endpoint eenvoudig via de id)
 - Delete: /delete/:id (Eenvoudig een ingezonden formulier verwijderen via id)


## Tests
Endpoints: Testen of de verschillend endpoints werken.

### Running the tests 
1.	Ga naar de api map
2.	Klik rechtermuisklik => open in “integrated terminal”
3.	Schrijf in de terminal “npm test”

## Problemen
Voor problemen of vragen wordt meer uitleg gegeven in de contributing file.

## Github Flow vs GIT Flow
Dit project maakt gebruik van Github flow. De reden waarom ik Github flow heb gekozen is omdat het 
bijna alle functionaliteiten biedt die Git Flow doet, zoals feature branches. Maar in tegenstelling tot Git Flow, combineert GitHub Flow de mainline en release branches tot een "master" en behandelt hotfixes net als feature branches. Dit vereenvoudigde model is beter geschikt voor modellen voor continue levering, waarbij wijzigingen snel en eenvoudig kunnen worden aangebracht, soms meerdere keren per dag. Exact wat mijn project dus nodig heeft.

## Status
Nog in development.
## Auteur
Kelly Peeters

CHANGELOG.md


# Changelog
Alle opmerkelijke wijzigingen aan dit project zullen in dit bestand gedocumenteerd worden.

## [1.0.0] - 2020-12-03
### Added
- Basis documenten toegevoegd
- Html toegevoegd
- Contributing, code-of-conduct en changelog toegevoegd
- Css toegevoegd

## [1.0.1] - 2020-12-04
### Added
- Docker toegevoegd 
- Ci pipline toegevoegd
- Jest toegevoegd

### Changed
- Mapstructuur aangepast
  
## [1.0.2] - 2020-12-25
### Fix
- Gegegevens kwijt door pull requests en push => Back-up teruggeplaatst

## [1.1.0] - 2020-12-26
### Added
- Get, update en delete actions toegevoegd

### Changed
- Css aangepast
- Contributing, readme file aangevuld
  
  ## [1.1.1] - 2021-01-01
### Changed
- Contributing file
- Readme file
- Database.test.js (connectie tests verwijderd)

 ## [1.1.2] - 2021-01-02

 ### Added
- Unit tests toegevoegd om email te testen

 ## [1.1.3] - 2021-01-07

 ### Added
- Endpoints failing tests toegevoegd

### Changed
- Endpoints notatie veranderd /get/vragenPerCategorie naar /getVragenPerCategorie:id

## [1.1.3] - 2021-01-07

 ### Added
- 2 get endpoints toegevoegd die data uit beide tables haalt
- Inline documentatie toegevoegd  
  
### Changed
- Tests verplaatst naar integration folder, zodat circli werkt

## [1.1.4] - 2021-01-17

 ### Added
- End-to-end-test toegevoegd

api/helper/helpers.test.js
  
let Helpers = require('../utils/helpers.js');

describe('test of de email geldig is', () => {
    test('email moet een string zijn', () => {
        expect(Helpers.checkValidEmail()).toBeFalsy();
        expect(Helpers.checkValidEmail(101)).toBeFalsy();
        expect(Helpers.checkValidEmail([])).toBeFalsy(); 
        expect(Helpers.checkValidEmail(null)).toBeFalsy();
        expect(Helpers.checkValidEmail(-56)).toBeFalsy();
    });

    test('email moet langer dan 1 zijn', () => {
        expect(Helpers.checkValidEmail("h")).toBeFalsy();
    });

    test('email moet korter zijn dan 254', () => {
        expect(Helpers.checkValidEmail("valid@email.com").length).toBeLessThanOrEqual(254);
        expect(Helpers.checkValidEmail("impggdcshvzxdbwphvjnschiazqbgldxpihnjfwamljsdeqvljlcaqileanjldgnljdtmxnfaodoukkxzmqiqgozhtnzmbicmcrjwrasocqmjdiqivdspimvelcumcphsaoayzxblkmlvwnbxaamlmzmpuvhhfklkfvtdghvgphizhfithfdcnkqslduwfgqfjnbwyqpnqiqfbbqgagryuwyhqevdpenunejrabtmatwzkdwdfgxklumliggotwisnnttm")).toBeFalsy();
    });

    test('email moet de juiste tekens bevatten', () => {
        expect(Helpers.checkValidEmail("/*")).toBeFalsy();
    });
});

api/__tests__/helpers.test.js
const Helpers = require ('../utils/helpers')

describe('helpers test',() => {
    test('check if function generates something', () => {
        expect(Helpers.generateUUID()).not.toBeUndefined();
    })
    test('check if generated is UUID', () => {
        expect(Helpers.generateUUID()).toMatch(/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/);
    })
})

describe('limit tests',() => {
    test('requires one argument', () => {
        expect(Helpers.limitToten()).toBeUndefined();
    })
    test('if only accepts arrays', () => {
        expect(Helpers.limitToten('string voorbeeld')).toBeUndefined();
        expect(Helpers.limitToten([])).not.toBeUndefined();
    })
    test('if only return 10 items', () => {
        expect(Helpers.limitToten([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]).length).toBe(10);
    })
})

describe('lengthCheck test', () => {
    test('de titel moet een string zijn', () => {
        expect(Helpers.checkTitleLength()).toBeFalsy();
        expect(Helpers.checkTitleLength(101)).toBeFalsy();
        expect(Helpers.checkTitleLength([])).toBeFalsy(); 
        expect(Helpers.checkTitleLength(null)).toBeFalsy();
        expect(Helpers.checkTitleLength(-56)).toBeFalsy();
    })
    test('checken dat de titel maximum 100 karakters heeft', () => {
        expect(Helpers.checkTitleLength("Hello world").length).toBeLessThanOrEqual(100);
        expect(Helpers.checkTitleLength("Dit is een te lange string die dus niet goedgekeurd zou moeten worden omdat het meer dan 100 karakters telt en daardoor kan het dus niet geaccepteerd worden in dit project")).toBeFalsy();
    })
})    


describe('opdracht delete checken',() => {
    test('check if function generates something', () => {
        expect(Helpers.generateUUID()).not.toBeUndefined();
    })
    test('check if de delete request alleen een UUID bevat', () => {
        expect(Helpers.generateUUID()).toMatch(/\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/);
    })
})

api/__tests__/database.test.js
const { Pool } = require('pg');
const supertest = require('supertest');
const app = require('./../server.js');

 const request = supertest(app);

describe('testing postgres', () => {

    let pgPool;

    beforeAll(() => {
        pgPool = new Pool({
            connectionString: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
        });
    });

    afterAll(async () => {
        await pgPool.end();
    });

    // test send to endpoint -> uuid is uitgekomen

    test('connection', async () => {
        const client = await pgPool.connect();
        try {
            await client.query('BEGIN');

            const { rows } = await client.query('SELECT * FROM storyblock');
            expect(rows).toBeInstanceOf(Array);

        } catch(err) {
          throw err;
        } finally {
            client.release();
        }
    }) 

    test('full connect', async (done) => {
        const client = await pgPool.connect();
        try {
            let uuid = null;
            await request.post('/storyblock') 
            .send({content: 'testing' })
            .expect(200)
            .then((res) => {
                uuid = res.body[0].uuid
                done()
        }).catch((e) => {
          console.log(e);
    })
    await client.query('BEGIN');
    const {rows} = await client.query(`SELECT * FROM storyblock WHERE uuid='${uuid}'`);
    expect(rows).toBeInstanceOf(Array);
    expect(rows.length).toBeGreaterThan(0);


} catch(err){
    throw err;
} finally {
    client.release();
}
    }) 
    test('full connect storyblock', async (done) => {
        const client = await pgPool.connect();
        try {
            let uuid = null;
            await request.post('/storyblock') 
            .send({ content: 'testing' })
            .expect(200)
            .then((res) => {
                uuid = res.body[0].uuid
                done()
        }).catch((e) => {
          console.log(e);
    })
    await client.query('BEGIN');
    const {rows} = await client.query(`SELECT * FROM storyblock WHERE uuid='$uuid'`);
    expect(rows).toBeInstanceOf(Array);
    expect(rows.length).toBeGreaterThan(0);


} catch(err){
    throw err;
} finally {
    client.release();
}
    }) 
})

/*describe('POST', () => {
    test('Should return single post after insert', async () => {
        const res = await request.post('/test')
        .send({ uuid: /\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/, content: 'test' });
  
      expect(res.status).toEqual(200);
      expect(res.body.data).toBeDefined();
      console.log('woohoow');
    });
  });*/

api/html/contact.js
$('#btnInsert').click((e) => {
  e.preventDefault();
});
    /*console.log('clicksubmit');
    const categoriesoort = document.getElementById('onderwerp').value;
    const voornaam = document.getElementById('fname').value;
    const achternaam = document.getElementById('lname').value;
    const email = document.getElementById('email').value;
    const bericht = document.getElementById('subject').value; 
    console.log(voornaam);

    let data = {
      categoriesoort: categoriesoort, 
      voornaam: voornaam, 
      achternaam: achternaam, 
      email: email, 
      bericht: bericht 
    };
    
    console.log(data);
  
    $.ajax({
      url: 'http://127.0.0.1:5432/post/formulier',
      method: 'POST',
      dataType: 'json',
      data: data, 
      timeout: 1000000
    }).done(function (data, status) {
      console.log(data);
      console.log(status);
    }).fail(function (er1, er2) {
      console.log(er1);
      console.log(er2);
    }); 
  });*/
 /*$.ajax({
      url: 'http://localhost:3000/get/:alleVragen',
      method: 'GET',  
      dataType: 'json',
      timeout: 1000000,
      cache: false
    }).done(function (data) {
      for (let b = 0; b < data.length; b++) {
        console.log(data);
        $('#IngeleverdeForms').append(`<p"> ${data[b].value} </p>`);
      }
    }).fail(function (er1, er2) {
      console.log(er1);  
      console.log(er2); 
    });
  }); */

  /*$.ajax({
    url: 'http://localhost:3000/get/:alleVragen',
    method: 'GET', 
    headers: {  'Access-Control-Allow-Origin': 'http://localhost:3000/get/:alleVragen' }, 
    dataType: 'json',
    timeout: 1000000
  }).done(function (data) {
      console.log(data);
  }).fail(function (er1, er2) {
    console.log(er1);
    console.log(er2); 
  });
});  */

api/helper/DatabaseHelper.js

let db;
try {
  db = require('knex')({
    client: 'pg',
    connection: {
      host: process.env.POSTGRES_HOST || '127.0.0.1',
      user: process.env.POSTGRES_USER || 'example',
      password: process.env.POSTGRES_PASSWORD || 'example',
      database: process.env.POSTGRES_DB || 'test'
    },
    searchPath: ['knex', 'public']
  });

  async function initialiseTables() {
    await db.schema.hasTable('records').then(async (exists) => {
      if (!exists) {
        await db.schema
          .createTable('records', (table) => {
            table.increments();
            table.uuid('uuid');
            table.string('question');
            table.string('answer');
            table.timestamps(true, true);
          })
          .then(async () => {
            console.log('created table records');
          });

      }
    })
      .catch((e) => {
        console.log(e);
      })
  }
  initialiseTables()
}
catch (e) {
  console.log("could not start up")
}


module.exports = db

api/src/server.js
const express = require('express')
const bodyParser = require('body-parser');
const http = require('http');
const {
  Pool
} = require("pg");

const port = 3000

// Connectie maken met de database
const pg = require('knex')({
  client: 'pg',
  version: '9.6',
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});

const client = new Pool({
  user: "example",
  host: "localhost",
  database: "test",
  password: "example",
  port: "5432"
});

const app = express();
http.Server(app);


app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    // to support URL-encoded bodies
    extended: true
  })
);

app.get('/test', (req, res) => {
  console.log("test");
  res.status(200).send();
});

/**
 * [Initialiseer tables als deze nog niet bestaan]
 * @param: /
 * @returns: /
 */
async function initialiseTables() {
  await pg.schema.hasTable('vragen').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('vragen', (table) => {
          table.increments();
          table.string('categoriesoort');
          table.string('voornaam');
          table.string('achternaam');
          table.string('email');
          table.string('bericht');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table vragen');
          for (let i = 0; i < 10; i++) {
            await pg.table('vragen').insert({
              categoriesoort,
              voornaam,
              achternaam,
              email,
              bericht,
              id: `random element number ${i}`
            })
          }
        });

    }
  });

  await pg.schema.hasTable('categorie').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('categorie', (table) => {
          table.increments();
          table.string('categoriesoort');
          table.string('summary');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table categorie');
        });
    }
  });
}

/**
 * [Get alle ingediende formulieren]
 * @param: /
 * @returns: Alle vragen (json, status 200 ok)
 */
app.get('/alleVragen', async (req, res) => {
  const result = await pg
    .select(['categoriesoort', 'voornaam', 'achternaam', 'email', 'bericht'])
    .from('vragen')
  //console.log("De vragen in de database zijn " + (JSON.stringify(result)));
  res.json({
    res: result
  })
})

/**
 * [Get alle ingediende formulieren van bepaalde categoriesoort]
 * @param: {string} categoriesoort
 * @returns: Alle vragen met bepaalde categoriesoort (json, status 200 ok) of error (status 400 bad request)
 */
app.get('/categorie/:categoriesoort', async (req, res) => {

  if (!req.body.categoriesoort) {
    return res.status(400).send({
      'message': 'Categorie is missing'
    });
  }

  const result = await pg
    .select(['voornaam', 'achternaam', 'email', 'bericht'])
    .from('vragen')
    .where({
      categoriesoort: req.body.categoriesoort
    })
  res.json({
    res: result
  })
  console.log(result);
})

/**
 * [Post een formulier]
 * @param: {string} categoriesoort
 * @param: {string} voornaam
 * @param: {string} achternaam
 * @param: {string} email
 * @param: {string} bericht
 * @returns: row toegevoegd in database (status: 200 ok) of error (status 400 bad request)
 */
app.post('/formulier', async (req, res) => {

  if (!req.body.categoriesoort || !req.body.voornaam || !req.body.achternaam || !req.body.email || !req.body.bericht) {
    return res.status(400).send({
      'message': 'Some values are missing'
    });
  }

  const {
    categoriesoort,
    voornaam,
    achternaam,
    email,
    bericht
  } = req.body

  client.query(
    'INSERT INTO vragen (categoriesoort, voornaam, achternaam, email, bericht) VALUES ($1, $2, $3, $4, $5)',
    [categoriesoort, voornaam, achternaam, email, bericht],
    (error) => {
      if (error) {
        throw error
      }
      console.log(categoriesoort, voornaam, achternaam, email, bericht);
      res.status(201).json({
        status: 'success',
        message: 'Insert goed gelukt'
      })
    },
  )
})

/**
 * [Get data van beide tables]
 * @param: /
 * @returns: get alle ingediende formulieren met categoriestoorten met data id, categoriesoort, tijd en bericht (status: 200 ok of status: 400 bad request)
 */
app.get('/join', async (req, res) => {
  await pg
    .from('vragen')
    .join('categorie', 'vragen.categoriesoort', 'categorie.categoriesoort')
    .select(['vragen.id', 'categorie.categoriesoort_id', 'categorie.categoriesoort', 'vragen.created_at', 'vragen.bericht'])
    .then(data => {
      res.json({
        res: data
      });
    })
});

/**
 * [Get data van beide tables]
 * @param: /
 * @returns: get alle ingediende formulieren met categoriestoorten met data id, categoriesoort, tijd en bericht (status: 200 ok of status: 400 bad request)
 */
app.get('/joinAlles', async (req, res) => {
  await pg
    .from('vragen')
    .join('categorie', 'vragen.categoriesoort', 'categorie.categoriesoort')
    .select('vragen.*', 'categorie.*')
    .then(data => {
      res.json({
        res: data
      });
    })
});

/**
 * [Delete een formulier]
 * @param: {number} id
 * @returns: row verwijderd uit database (status: 200 ok) of (status: 400 bad request)
 */
app.delete('/deleteVraag/:id', async (req, res) => {

  if (!req.body.id) {
    return res.status(400).send({
      'message': 'Id is missing'
    });
  }

  client.query(
    `DELETE FROM vragen WHERE id=$1`,
    [req.body.id],
    (error) => {
      if (error) {
        throw error
      }
      console.log("deleted row with id = " + req.body.id);
      res.status(200).json({
        status: 'success',
        message: 'delete goed gelukt'
      })
    },
  )
});

/**
 * [Update een formulier]
 * @param: {number} id
 * @returns: row geupdate in database (status: 200 OK) of (status: 400 bad request)
 */
app.patch("/updateVraag/:id", (req, res) => {

  if (!req.body.id) {
    return res.status(400).send({
      'message': 'Id is missing'
    });
  }

  client.query(
    `UPDATE vragen SET categoriesoort = $1, voornaam = $2, achternaam = $3, email = $4, bericht = $5, updated_at = CURRENT_TIMESTAMP WHERE id = $6`,
    [req.body.categoriesoort, req.body.voornaam, req.body.achternaam, req.body.email, req.body.bericht, req.body.id],
    (error) => {
      if (error) {
        throw error
      }
      console.log("updated row with id = " + req.body.id);
      res.status(200).json({
        status: 'success',
        message: 'update goed gelukt'
      })
    },
  )
});

initialiseTables()

module.exports = app;